File: problem_004.py
#!/usr/bin/env python3
import time

def blurb():
  print("""

  A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

  Find the largest palindrome made from the product of two 3-digit numbers.

  """)

def main():
  start = time.time()

  LOWER_LIMIT = 10000
  UPPER_LIMIT = 998001

  NUMBERS = list( range(UPPER_LIMIT, LOWER_LIMIT, -1) )
  PALINDROMES = []

  for number in NUMBERS:
    number = str(number)
    if number == number[::-1]:
      PALINDROMES.append( int(number) )
  
  sorted(PALINDROMES)

  THREE_DIGIT_PALINDROMES = set()

  for factor in range(999,100,-1):
    for palindrome in PALINDROMES[::-1]:
      if palindrome % factor == 0 and 10 > int((palindrome / factor) / 100) > 1 :
          THREE_DIGIT_PALINDROMES.add(palindrome)

  RESULT = max(THREE_DIGIT_PALINDROMES)

  print(RESULT)

  end   = time.time() - start
  print("Completed in %s" % end)

if __name__ == '__main__':
  main()

File: go/problem004.go
package main


import (
    "fmt"
    "time"
    "strconv"
)


func blurb() {
    fmt.Println(`
    A palindromic number reads the same both ways. The largest palindrome made
    from the product of two 2-digit numbers is 9009 = 91 × 99.

    Find the largest palindrome made from the product of two 3-digit numbers.
    `)
}


func Reverse(s string) string {
    runes := []rune(s)
    for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
        runes[i], runes[j] = runes[j], runes[i]
    }
    return string(runes)
}


func findPalindromicNumbers(lowerBound, upperBound int) []int {
    palindromes := make([]int, 0)
    for i := lowerBound; i < upperBound ; i++ {
        iString := strconv.Itoa(i)
        if iString == Reverse(iString) {
            palindromes = append(palindromes, i)
        }
    }
    return palindromes
}


func findLargestFactor(numbers []int, lowerBound int, upperBound int) int {
    var result int
    for i := len(numbers) - 1 ; i >= 0 ; i-- {
        var palindromicNumber int = numbers[i]
        for j := upperBound ; j >= lowerBound ; j-- {
            if palindromicNumber % j == 0 && lowerBound <= j && j <= upperBound {
                var divisor int = palindromicNumber / j
                if lowerBound <= divisor && divisor <= upperBound {
                    fmt.Printf("%d * %d == %d\n", j, divisor, palindromicNumber)
                    return palindromicNumber
                }
            }
        }
    }
    return result
}


func problem004() (result int) {
    palindromicNumbers := findPalindromicNumbers(100*100, 999*999)
    result = findLargestFactor(palindromicNumbers, 100, 999)
    return
}


func main() {
    blurb()
    var start time.Time = time.Now()
    var result int = problem004()
    var end time.Time = time.Now()
    var elapsed time.Duration = end.Sub(start)
    fmt.Printf("Result: %v\n", result)
    fmt.Printf("Completed in: %v\n", elapsed)
}
